File: test/java/brs/web/api/http/handler/SellAliasTest.java
package brs.web.api.http.handler;

import brs.*;
import brs.common.QuickMocker;
import brs.common.QuickMocker.MockParam;
import brs.fluxcapacitor.FluxCapacitor;
import brs.fluxcapacitor.FluxValues;
import brs.services.ParameterService;
import brs.web.api.http.common.APITransactionManager;
import brs.web.api.http.common.ParameterException;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import javax.servlet.http.HttpServletRequest;

import static brs.Constants.MAX_BALANCE_NQT;
import static brs.TransactionType.Messaging.ALIAS_SELL;
import static brs.web.api.http.common.JSONResponses.*;
import static brs.web.api.http.common.Parameters.PRICE_NQT_PARAMETER;
import static brs.web.api.http.common.Parameters.RECIPIENT_PARAMETER;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static org.powermock.api.mockito.PowerMockito.mockStatic;

@RunWith(PowerMockRunner.class)
@PrepareForTest(Signum.class)
public class SellAliasTest extends AbstractTransactionTest {

  private SellAlias t;

  private ParameterService parameterServiceMock;
  private Blockchain blockchainMock;
  private APITransactionManager apiTransactionManagerMock;

  @Before
  public void setUp() {
    parameterServiceMock = mock(ParameterService.class);
    blockchainMock = mock(Blockchain.class);
    apiTransactionManagerMock = mock(APITransactionManager.class);

    t = new SellAlias(parameterServiceMock, blockchainMock, apiTransactionManagerMock);
  }

  @Test
  public void processRequest() throws SignumException {
    final int priceParameter = 10;
    final int recipientId = 5;

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(PRICE_NQT_PARAMETER, priceParameter),
        new MockParam(RECIPIENT_PARAMETER, recipientId)
    );

    final long aliasAccountId = 1L;
    final Alias mockAlias = mock(Alias.class);
    when(mockAlias.getAccountId()).thenReturn(aliasAccountId);

    final Account mockSender = mock(Account.class);
    when(mockSender.getId()).thenReturn(aliasAccountId);

    when(parameterServiceMock.getSenderAccount(req)).thenReturn(mockSender);
    when(parameterServiceMock.getAlias(req)).thenReturn(mockAlias);

    mockStatic(Signum.class);
    final FluxCapacitor fluxCapacitor = QuickMocker.fluxCapacitorEnabledFunctionalities(FluxValues.DIGITAL_GOODS_STORE);
    when(Signum.getFluxCapacitor()).thenReturn(fluxCapacitor);
    doReturn(Constants.FEE_QUANT_SIP3).when(fluxCapacitor).getValue(eq(FluxValues.FEE_QUANT));

    final Attachment.MessagingAliasSell attachment = (Attachment.MessagingAliasSell) attachmentCreatedTransaction(() -> t.processRequest(req), apiTransactionManagerMock);
    assertNotNull(attachment);

    assertEquals(ALIAS_SELL, attachment.getTransactionType());
    assertEquals(priceParameter, attachment.getPriceNqt());
  }

  @Test
  public void processRequest_missingPrice() throws SignumException {
    final HttpServletRequest req = QuickMocker.httpServletRequest();

    assertEquals(MISSING_PRICE, t.processRequest(req));
  }

  @Test
  public void processRequest_incorrectPrice_unParsable() throws SignumException {
    final HttpServletRequest req = QuickMocker.httpServletRequest(
      new MockParam(PRICE_NQT_PARAMETER, "unParsable")
    );

    assertEquals(INCORRECT_PRICE, t.processRequest(req));
  }

  @Test(expected = ParameterException.class)
  public void processRequest_incorrectPrice_negative() throws SignumException {
    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(PRICE_NQT_PARAMETER, -10)
    );

    t.processRequest(req);
  }

  @Test(expected = ParameterException.class)
  public void processRequest_incorrectPrice_overMaxBalance() throws SignumException {
    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(PRICE_NQT_PARAMETER, MAX_BALANCE_NQT + 1)
    );

    t.processRequest(req);
  }

  @Test
  public void processRequest_incorrectRecipient_unparsable() throws SignumException {
    final int price = 10;

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(PRICE_NQT_PARAMETER, price),
        new MockParam(RECIPIENT_PARAMETER, "unParsable")
    );

    assertEquals(INCORRECT_RECIPIENT, t.processRequest(req));
  }

  @Test
  public void processRequest_incorrectRecipient_zero() throws SignumException {
    final int price = 10;
    final int recipientId = 0;

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(PRICE_NQT_PARAMETER, price),
        new MockParam(RECIPIENT_PARAMETER, recipientId)
    );

    assertEquals(INCORRECT_RECIPIENT, t.processRequest(req));
  }

  @Test
  public void processRequest_incorrectAliasOwner() throws SignumException {
    final int price = 10;
    final int recipientId = 5;

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(PRICE_NQT_PARAMETER, price),
        new MockParam(RECIPIENT_PARAMETER, recipientId)
    );

    final long aliasAccountId = 1L;
    final Alias mockAlias = mock(Alias.class);
    when(mockAlias.getAccountId()).thenReturn(aliasAccountId);

    final long mockSenderId = 2l;
    final Account mockSender = mock(Account.class);
    when(mockSender.getId()).thenReturn(mockSenderId);

    when(parameterServiceMock.getSenderAccount(req)).thenReturn(mockSender);
    when(parameterServiceMock.getAlias(req)).thenReturn(mockAlias);

    assertEquals(INCORRECT_ALIAS_OWNER, t.processRequest(req));
  }

}

File: src/test/kotlin/brs/http/SellAliasTest.kt
package brs.http

import brs.*
import brs.Constants.MAX_BALANCE_NQT
import brs.common.QuickMocker
import brs.common.QuickMocker.MockParam
import brs.fluxcapacitor.FluxValues
import brs.http.JSONResponses.INCORRECT_ALIAS_OWNER
import brs.http.JSONResponses.INCORRECT_PRICE
import brs.http.JSONResponses.INCORRECT_RECIPIENT
import brs.http.JSONResponses.MISSING_PRICE
import brs.http.common.Parameters.PRICE_NQT_PARAMETER
import brs.http.common.Parameters.RECIPIENT_PARAMETER
import brs.services.ParameterService
import brs.transaction.TransactionType
import brs.transaction.messaging.AliasSell
import com.nhaarman.mockitokotlin2.doReturn
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

@RunWith(JUnit4::class)
class SellAliasTest : AbstractTransactionTest() {

    private lateinit var t: SellAlias
    private lateinit var dp: DependencyProvider
    private lateinit var parameterServiceMock: ParameterService
    private lateinit var blockchainMock: Blockchain
    private lateinit var apiTransactionManagerMock: APITransactionManager

    @Before
    fun setUp() {
        parameterServiceMock = mock()
        blockchainMock = mock()
        apiTransactionManagerMock = mock()
        dp = QuickMocker.dependencyProvider(parameterServiceMock, blockchainMock, apiTransactionManagerMock)
        t = SellAlias(dp)
    }

    @Test
    fun processRequest() {
        val priceParameter = 10
        val recipientId = 5

        val request = QuickMocker.httpServletRequest(
                MockParam(PRICE_NQT_PARAMETER, priceParameter),
                MockParam(RECIPIENT_PARAMETER, recipientId)
        )

        val aliasAccountId = 1L
        val mockAlias = mock<Alias>()
        whenever(mockAlias.accountId).doReturn(aliasAccountId)
        whenever(mockAlias.aliasName).doReturn("")

        val mockSender = mock<Account>()
        whenever(mockSender.id).doReturn(aliasAccountId)

        whenever(parameterServiceMock.getSenderAccount(request)).doReturn(mockSender)
        whenever(parameterServiceMock.getAlias(request)).doReturn(mockAlias)

        dp.fluxCapacitor = QuickMocker.fluxCapacitorEnabledFunctionalities(FluxValues.DIGITAL_GOODS_STORE)
        dp.transactionTypes = TransactionType.getTransactionTypes(dp)

        val attachment = attachmentCreatedTransaction({ t.processRequest(request) }, apiTransactionManagerMock) as Attachment.MessagingAliasSell
        assertNotNull(attachment)

        assertTrue(attachment.transactionType is AliasSell)
        assertEquals(priceParameter.toLong(), attachment.priceNQT)
    }

    @Test
    fun processRequest_missingPrice() {
        val request = QuickMocker.httpServletRequest()

        assertEquals(MISSING_PRICE, t.processRequest(request))
    }

    @Test
    fun processRequest_incorrectPrice_unParsable() {
        val request = QuickMocker.httpServletRequest(
                MockParam(PRICE_NQT_PARAMETER, "unParsable")
        )

        assertEquals(INCORRECT_PRICE, t.processRequest(request))
    }

    @Test(expected = ParameterException::class)
    fun processRequest_incorrectPrice_negative() {
        val request = QuickMocker.httpServletRequest(
                MockParam(PRICE_NQT_PARAMETER, -10L)
        )

        t.processRequest(request)
    }

    @Test(expected = ParameterException::class)
    fun processRequest_incorrectPrice_overMaxBalance() {
        val request = QuickMocker.httpServletRequest(
                MockParam(PRICE_NQT_PARAMETER, MAX_BALANCE_NQT + 1)
        )

        t.processRequest(request)
    }

    @Test
    fun processRequest_incorrectRecipient_unparsable() {
        val price = 10

        val request = QuickMocker.httpServletRequest(
                MockParam(PRICE_NQT_PARAMETER, price),
                MockParam(RECIPIENT_PARAMETER, "unParsable")
        )

        assertEquals(INCORRECT_RECIPIENT, t.processRequest(request))
    }

    @Test
    fun processRequest_incorrectRecipient_zero() {
        val price = 10
        val recipientId = 0

        val request = QuickMocker.httpServletRequest(
                MockParam(PRICE_NQT_PARAMETER, price),
                MockParam(RECIPIENT_PARAMETER, recipientId)
        )

        assertEquals(INCORRECT_RECIPIENT, t.processRequest(request))
    }

    @Test
    fun processRequest_incorrectAliasOwner() {
        val price = 10
        val recipientId = 5

        val request = QuickMocker.httpServletRequest(
                MockParam(PRICE_NQT_PARAMETER, price),
                MockParam(RECIPIENT_PARAMETER, recipientId)
        )

        val aliasAccountId = 1L
        val mockAlias = mock<Alias>()
        whenever(mockAlias.accountId).doReturn(aliasAccountId)

        val mockSenderId = 2L
        val mockSender = mock<Account>()
        whenever(mockSender.id).doReturn(mockSenderId)

        whenever(parameterServiceMock.getSenderAccount(request)).doReturn(mockSender)
        whenever(parameterServiceMock.getAlias(request)).doReturn(mockAlias)

        assertEquals(INCORRECT_ALIAS_OWNER, t.processRequest(request))
    }

}
